crates/extensions/tedge_mqtt_ext/src/test_helpers.rs

```diff
@@ -48,14 +33,88 @@ pub async fn assert_received_includes_json<I, S>(
 {
     for expected_msg in expected.into_iter() {
         let message = messages.recv().await.expect("MQTT channel closed");
-        assert_eq!(message.topic, Topic::new_unchecked(expected_msg.0.as_ref()));
-        let payload = serde_json::from_str::<serde_json::Value>(
-            message.payload_str().expect("non UTF-8 payload"),
-        )
-        .expect("non JSON payload");
-        assert_json_include!(
-            actual: payload,
-            expected: expected_msg.1
-        );
+        assert_message_includes_json(&message, expected_msg);
+    }
+}
+
+pub fn assert_messages_contains_str<I, S>(messages: &mut Vec<MqttMessage>, expected: I)
+where
+    I: IntoIterator<Item = (S, S)>,
+    S: AsRef<str> + Debug,
+{
+    let mut expected = expected.into_iter();
+    loop {
+        match (messages.is_empty(), expected.next()) {
+            (false, Some(expected_msg)) => {
+                let message = messages.remove(0);
+                assert_message_contains_str(&message, expected_msg);
+            }
+            (true, Some(expected_msg)) => {
+                panic!(
+                    "Input messages exhausted while expecting: {:?}",
+                    expected_msg
+                )
+            }
+            _ => break,
+        }
     }
 }
+
+pub fn assert_messages_includes_json<I, S>(messages: &mut Vec<MqttMessage>, expected: I)
+where
+    I: IntoIterator<Item = (S, serde_json::Value)>,
+    S: AsRef<str>,
+{
+    let mut expected = expected.into_iter();
+    loop {
+        match (messages.is_empty(), expected.next()) {
+            (false, Some(expected_msg)) => {
+                let message = messages.remove(0);
+                assert_message_includes_json(&message, expected_msg);
+            }
+            (true, Some(expected_msg)) => {
+                panic!(
+                    "Input messages exhausted while expecting message on topic: {:?} with payload: {:?}",
+                    expected_msg.0.as_ref(), expected_msg.1
+                )
+            }
+            _ => break,
+        }
+    }
+}
+
+pub fn assert_message_contains_str<S>(message: &Message, expected: (S, S))
+where
+    S: AsRef<str> + Debug,
+{
+    let expected_topic = expected.0.as_ref();
+    let expected_payload = expected.1.as_ref();
+    assert_eq!(
+        message.topic,
+        Topic::new_unchecked(expected_topic),
+        "\nReceived unexpected message: {:?}",
+        message
+    );
+    let payload = message.payload_str().expect("non UTF-8 payload");
+    assert!(
+        payload.contains(expected_payload),
+        "Payload assertion failed.\n Actual: {} \n Expected: {}",
+        payload,
+        expected_payload
+    )
+}
+
+pub fn assert_message_includes_json<S>(message: &Message, expected: (S, serde_json::Value))
```

Comment:
Is there relevance to the use of `contains` in the `str` functions, but `includes` in the `json` functions? If not, could one of them be changed to match the other?

Reply:
We're doing a loose assertion with just `contains` for the `str` values as the target values were mostly SmartREST messages and we dint want to force the user to always provide the whole value, with trailing `\n` or other dynamic fields like URLs, timestamps etc. Similar reason for `includes` in `json` as well, as we often want to skip the `time` field in them. We could add an extra `exact` boolean parameter to this function for strict matching. But, we've never really come across one.

> If not, could one of them be changed to match the other?

I didn't get what you're suggesting here.

Reply:
> I didn't get what you're suggesting here.

In my mind, "contains" and "includes" are essentially synonymous here, and if they are used to mean the same thing (or equivalent things as applied to strings and json objects), I'd prefer if both functions used the same word. In this context, having two different but similar words generally implies that there is some important distinction to be made between the two (or at least it does in my mind).

Reply:
Aah okay. Got it. The new name of the combined API is just `assert_messages_matching`, which I hope is better.